File: ios/Classes/InteractionModel.swift
//
//  InteractionModel.swift
//  xyo_ble
//
//  Created by Darren Sutherland on 6/27/19.
//

import Foundation
import secp256k1
import sdk_core_swift
import sdk_bletcpbridge_swift
import sdk_objectmodel_swift

class InteractionModel {
    private var byteHash: [UInt8]
    private var boundWitness: XyoBoundWitness?

    /// this should never change
    fileprivate(set) lazy var hash: String = self.byteHash.toBase58String()

    /// this shpuld never change
    let date: Date

    /// this should never change
    let linked: Bool

    var huerestics : [String: String] {
        guard let bw = self.boundWitness else {
            return [:]
        }

        return XyoHumanHeuristics.getHumanHeuristics(boundWitness: bw)
    }

    var bytes: String {
        guard let bw = self.boundWitness else {
            return NSLocalizedString("Bound witness has been offloaded", comment: "bw offloaded interactions view model")
        }

        return bw.getBuffer().toByteArray().toHexString()
    }

    var humanName: String {
        guard let bw = self.boundWitness else {
            return NSLocalizedString("Bound witness has been offloaded", comment: "bw offloaded interactions view model")
        }

        return XyoHumanName.getHumanName(boundWitmess: bw, publicKey: BridgeManager.instance.getPrimaryPublicKey())
    }

    var parties: [String] {
        var publicKeys = [String]()
        guard let bw = self.boundWitness else {
            return publicKeys
        }

        for byteKey in getPrimaryPublicKeysFromBoundWitness(boundWitness: bw) {
            publicKeys.append(byteKey?.toBase58String() ?? NSLocalizedString("No public key", comment: "no public key interactions view model"))
        }

        return publicKeys
    }

    private func getPrimaryPublicKeysFromBoundWitness (boundWitness: XyoBoundWitness) -> [[UInt8]?] {
        var publicKeys = [[UInt8]?]()

        do {
            let numberOfParties = try boundWitness.getNumberOfParties() ?? 0

            if numberOfParties < 1 {
                return []
            }

            for i in 0...numberOfParties - 1 {
                publicKeys.append(try getPublicKeyFromPartyIndex(boundWitness: boundWitness, i: i))
            }

        } catch {
            return []
        }

        return publicKeys
    }

    private func getPublicKeyFromPartyIndex (boundWitness: XyoBoundWitness, i: Int) throws -> [UInt8]? {
        guard let fetterOfParty = try boundWitness.getFetterOfParty(partyIndex: i) else {
            return nil
        }

        guard let publicKeySet = try fetterOfParty.get(id: XyoSchemas.KEY_SET.id).first as? XyoIterableStructure else {
            return nil
        }

        if try publicKeySet.getCount() < 1 {
            return nil
        }

        return try publicKeySet.get(index: 0).getBuffer().toByteArray()
    }

    init(_ hash: [UInt8], date: Date, linked: Bool = true) {
        self.byteHash = hash
        self.date = date
        self.linked = linked
        do {
            boundWitness = try BridgeManager.instance.blockRepo.getOriginBlock(originBlockHash: self.byteHash)
        } catch {}
    }

    var toBuffer: DeviceBoundWitness {
        return DeviceBoundWitness.with {
            $0.bytes = self.bytes
            $0.byteHash = self.byteHash.toHexString()
            $0.linked = self.linked
            $0.huerestics = self.huerestics
            $0.parties = self.parties
            $0.humanName = self.humanName
        }
    }

}

File: android/src/main/kotlin/network/xyo/ble/xyo_ble/InteractionModel.kt
package network.xyo.ble.xyo_ble

import kotlinx.coroutines.runBlocking
import network.xyo.ble.flutter.protobuf.BoundWitness
import network.xyo.sdkcorekotlin.boundWitness.XyoBoundWitness
import network.xyo.sdkcorekotlin.repositories.XyoOriginBlockRepository
import network.xyo.sdkcorekotlin.schemas.XyoSchemas
import network.xyo.sdkobjectmodelkotlin.structure.XyoIterableStructure
import network.xyo.sdkobjectmodelkotlin.toHexString
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.LinkedHashMap

@kotlin.ExperimentalUnsignedTypes
class InteractionModel {
    private var byteHash = ByteArray(0)
    private var boundWitness: XyoBoundWitness? = null

    /// this should never change
    var hash: String = this.byteHash.toBase58String()

    /// this shpuld never change
    var date: Date

    /// this should never change
    var linked: Boolean

    constructor(blockRepo: XyoOriginBlockRepository, hash: ByteArray, date: Date, linked: Boolean = true) {
        this.byteHash = hash
        this.date = date
        this.linked = linked

        runBlocking {
            boundWitness = blockRepo.getOriginBlockByBlockHash(hash).await()
        }
    }

    constructor(block: XyoBoundWitness?, hash: ByteArray, date: Date, linked: Boolean = true) {
        this.byteHash = hash
        this.date = date
        this.linked = linked
        this.boundWitness = block
    }


    fun getHuerestics(): LinkedHashMap<String, String> {
        val bw = this.boundWitness
        if (bw != null) {
            return HumanReadableBoundWitness.getAllHumanHuerestics(bw)
        }

        return LinkedHashMap()
    }

    fun getBytes(): String { 
        val bw = this.boundWitness

        if (bw != null) {
            return bw.bytesCopy.toHexString()
        }

        return "Bound witness has been offloaded"
    }

    fun getHumanName(publicKey: ByteArray?): String {
        val bw = this.boundWitness

        if (bw != null) {
            return HumanReadableBoundWitness.getHumanBoundWitnessName(bw, publicKey)
        }

        return "Bound witness has been offloaded"
    }

    fun parties(): Array<String> {
        val publicKeys = ArrayList<String>()
        val bw = this.boundWitness
        if (bw != null) {
            for (byteKey in getPrimaryPublicKeysFromBoundWitness(bw)) {
                publicKeys.add(byteKey.toBase58String())
            }
        }

        return publicKeys.toTypedArray()
    }

    private fun getPrimaryPublicKeysFromBoundWitness (boundWitness: XyoBoundWitness): Array<ByteArray> {
        val publicKeys = ArrayList<ByteArray>()

        val numberOfParties = boundWitness.numberOfParties ?: 0

        if (numberOfParties < 1) {
            return arrayOf()
        }


        for (i in 0 until numberOfParties) {
            val key = getPublicKeyFromPartyIndex(boundWitness, i)

            if (key != null) {
                publicKeys.add(key)

            }
        }

        return publicKeys.toTypedArray()
    }

    private fun getPublicKeyFromPartyIndex (boundWitness: XyoBoundWitness, i: Int): ByteArray? {
        val fetterOfParty = boundWitness.getFetterOfParty(i) ?: return null
        val publicKeySet = fetterOfParty[XyoSchemas.KEY_SET.id].firstOrNull() as? XyoIterableStructure ?: return null


        if (publicKeySet.count < 1) {
            return null
        }

        return publicKeySet[0].bytesCopy
    }

    fun toBuffer(): BoundWitness.DeviceBoundWitness {
        val bytes: String = boundWitness?.bytesCopy?.toHexString() ?: String()
        val byteHash: String = hash
        val humanName: String = getHumanName(null)
        val huerestics: Map<String, String> = getHuerestics().toMap()
        val parties: Array<String> = parties()
        val linked = this.linked // not sure what this is
        //val unknownFields: ByteArray = byteArrayOf() // not sure what this is

        val builder = BoundWitness.DeviceBoundWitness.newBuilder()
        builder.bytes = bytes
        builder.byteHash = byteHash
        builder.humanName = humanName
        builder.linked = linked
        builder.addAllParties(parties.toList())
        builder.putAllHuerestics(huerestics)


        return builder.build()
    }
}
